File: Graph/Python/Count Servers that Communicate.py
# Problem link: https://leetcode.com/problems/count-servers-that-communicate/description/
# Time: O(n * m)
# space: O(n + m)
class Solution:
    def countServers(self, grid: List[List[int]]) -> int:
        rows, cols, result = len(grid), len(grid[0]), 0
        rowServers, colServers = [0] * rows, [0] * cols

        for row in range(rows):
            for col in range(cols):
                if grid[row][col]:
                    rowServers[row] += 1
                    colServers[col] += 1
        
        for row in range(rows):
            for col in range(cols):
                if grid[row][col] and (rowServers[row] > 1 or colServers[col] > 1):
                    result += 1
        
        return result

File: Graph/Java/Count Servers that Communicate.java
// Problem link: https://leetcode.com/problems/count-servers-that-communicate/description/
// Time: O(n * m)
// Space: O(n + m)
class Solution {
    public int countServers(int[][] grid) {
        int rows = grid.length, cols = grid[0].length, result = 0;
        int[] rowServers = new int[rows], colServers = new int[cols];

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(grid[row][col] == 1) {
                    rowServers[row]++;
                    colServers[col]++;
                }
            }
        }

        for(int row = 0; row < rows; row++) {
            for(int col = 0; col < cols; col++) {
                if(grid[row][col] == 1 && (rowServers[row] > 1 || colServers[col] > 1)) {
                    result++;
                }
            }
        }

        return result;
    }
}
